### ISSUE TITLE
Alternator: missing TableSizeBytes field in return of DescribeTable

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#7551]
The return from `DescribeTable` is missing the `TableSizeBytes` field. According to DynamoDB documentation, `TableSizeBytes` should have:

> The total size of the specified table, in bytes. DynamoDB updates this value approximately every six hours. Recent changes might not be reflected in this value. 

This estimate is available to us separately for each node, but the question is how do we sum it up for all nodes. Do we do this during the query, or periodically? What to do when some of the nodes are down?

We have an xfailing trivial test for this feature: `test_describe_table.py`::`test_describe_table_size`.

**Comment by nyh** [comment]
I had assumed that `TableSizeBytes` is the same size used by Amazon to bill the _storage_ of the table. However, it turns out that https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/CapacityUnitCalculations.html explains that for billing purposes, "_All items in DynamoDB require 100 bytes of storage overhead for indexing._", but https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithTables.Basics.html#WorkingWithTables.Basics.DescribeTable claims that "_This extra 100 bytes per item is not used ... by the DescribeTable operation._".

So we'll need to do some testing on AWS to better understand what `TableSizeBytes` really measures.

**Comment by ptrsmrn** [comment]
@mykaul how does `Field-Tier` label translate to the `Px` labels we use for prioritizing? I'm trying to understand how urgent is it.

**Comment by pdbossman** [comment]
> I had assumed that `TableSizeBytes` is the same size used by Amazon to bill the _storage_ of the table. However, it turns out that https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/CapacityUnitCalculations.html explains that for billing purposes,
> 
> "_All items in DynamoDB require 100 bytes of storage overhead for indexing._"
> 
> But https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithTables.Basics.html#WorkingWithTables.Basics.DescribeTable claims that:
> 
> "_This extra 100 bytes per item is not used ... by the DescribeTable operation._".
> 
> So we'll need to do some testing on AWS to better understand what `TableSizeBytes` really measures.

I believe what customers would want this field for is to know the size of a table on disk.  Every customer interaction I've had on table size - the customer is trying to find the largest space on disk as it's causing a storage crunch, and they would like to attempt to address that via deletes or more enhanced compression, and they'd want to see the size be reduced.

IMO if the TableSizeBytes does not reflect the ... space used by the table, it seems like a bug and degrades usefulness.

**Comment by nyh** [comment]
@pdbossman clearly the TableSizeBytes should "reflect" the space used by the table, but it's not obvious whether it should take into account replication (if RF=3, should we returned the total disk used or one third of it),  compression, indexing, and so on. I agree that we can decide to calculate this size in a different way than DynamoDB does - e.g., we can decide that if we PutItem 1000 items of size 1000 each and DynamoDB returns TableSizeBytes=1,000,000, it's fine that we return 1,777,777 because it's the number of bytes we really stored on disk. But we'll need to document this difference and explain why it's-not-a-bug-it's-a-feature.

But before we rush to do this (as "field-tier1" seems to suggestion), don't we already have in our monitoring and/or nodetool mechanisms for getting the on-disk size, which our CQL users are already using? Can't Alternator users use the same mechanisms?

**Comment by pdbossman** [comment]
This is open source repo, so yes those things are available for self-hosted clusters.
Managed services may not allow customers to use things like nodetool and have table level monitoring metrics.

**Comment by ptrsmrn** [comment]
Do we need a design here? I don't think so, another pair of eyes/some additional input as for how to make these calculations maybe?

**Comment by radoslawcybulski** [comment]
I'd say it should return disk usage (total - ie all nodes added), so essentially sum of sizes of SSTable's on disk.
Whom to ask to confirm this (or reject)? Should we add index files as well?

**Comment by nyh** [comment]
> I'd say it should return disk usage (total - ie all nodes added), so essentially sum of sizes of SSTable's on disk. Whom to ask to confirm this (or reject)?

@pdbossman agreed above (https://github.com/scylladb/scylladb/issues/7551#issuecomment-2457426156) and also I agreed (https://github.com/scylladb/scylladb/issues/7551#issuecomment-2457471365), that it is probably fine to use the sstable size on disk even if the result isn't exactly the same number as what DynamoDB returns. But we will need to document this exactly in compatibility.md - in case one of our users needs to know what this number measures.

Note that a single shard will need to answer this request, so it will need to know how much disk space all other shards and nodes are using. We're probably doing something about this for "nodetool tablestats", but if we didn't we could have each node periodically sum its disk usage to its gossip structure, and all nodes will learn about all other nodes via gossip. This would also solve the problem you asked about in https://github.com/scylladb/scylladb/issues/11466#issuecomment-2958833338 - if a node is temporarily down, its gossip data is still available to other nodes.

> Should we add index files as well?

Probably not because there's a separate count for index size (https://github.com/scylladb/scylladb/issues/11466). However, this is not something clearly explained in the documentation, so we should test in practice instead of guessing: Write a test that creates a table on DynamoDB, gets its TableSizeBytes, adds a GSI and gets TableSizeBytes again, to see if it grew larger. You'll may need to add very long sleeps (more than 6 hours) between each step because this is what the documentation says it takes for these counts to get updated. When you know the answer, you should document it in the same place we will document how our TableSizeBytes works.
